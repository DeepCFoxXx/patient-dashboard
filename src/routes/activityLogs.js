import express from "express";
import ActivityLog from "../models/ActivityLog.js";

const router = express.Router();

router.get("/", async (req, res) => {
  const logs = await ActivityLog.find().populate("patientId");
  res.json(logs);
});

router.post("/", async (req, res) => {
  try {
    const newLog = new ActivityLog(req.body);
    await newLog.save();
    res.status(201).json(newLog);
  } catch (err) {
    res.status(400).json({ error: err.message });
  }
});

export default router;
