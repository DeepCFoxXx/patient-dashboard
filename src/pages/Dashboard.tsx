import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import BarChart from "../components/BarChart";
import LineChart from "../components/LineChart";
import PieChart from "../components/PieChart";
import { activityLogs } from "../data/activityLogs";
import { patientData } from "../data/patientData";
import { calculateAverages, convertSecondsToHMS, prepareChartData, prepareRpeData, processSessionData } from "../helpers/sessionHelpers";
import "../styles/dashboard-styles.css";
import { ActivityLog, Patient, SessionDataMap } from "../types/globalTypes";

const Dashboard: React.FC = () => {
  const navigate = useNavigate();
  const [patient, setPatient] = useState<Patient | null>(null);
  const [logs, setLogs] = useState<ActivityLog[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    try {
      setPatient(patientData);
      setLogs(activityLogs);
    } catch {
      setError("Failed to load data");
    } finally {
      setLoading(false);
    }
  }, []);

  const sessionData: SessionDataMap = processSessionData(logs);
  const { averageDuration, averageReps, averageRpe } = calculateAverages(sessionData);
  const formattedAverageDuration = convertSecondsToHMS(averageDuration);
  const chartData = prepareChartData(logs);
  const rpeChartData = prepareRpeData(logs);

  const handleLogout = () => {
    localStorage.removeItem('token');
    navigate('/');
  };

  if (loading) return <p className="loading">Loading...</p>;
  if (error) return <p className="error">{error}</p>;

  const durationPieData = [
    { id: "Short", label: "Short", value: 10 },
    { id: "Medium", label: "Medium", value: 30 },
    { id: "Long", label: "Long", value: 20 },
  ];

  return (
    <div className="dashboard">
      <div className="card patient-details">
        <h2 className="underline">Patient Details</h2>
        {patient && (
          <div className="patient-info">
            <div className="patient-info-left">
              <p>Name: {patient.firstName} {patient.lastName}</p>
              <p>Date of Birth: {patient.dateOfBirth}</p>
              <p>Average Reps Per Session: {Math.round(averageReps)} reps</p>
              <p>Average Therapy Duration: {formattedAverageDuration}</p>
              <p>Average RPE Per Session: {averageRpe.toFixed(2)}</p>
            </div>
            <div className="patient-info-right">
              <h3>Average Therapy Duration</h3>
              <PieChart data={durationPieData} />
            </div>
          </div>
        )}
      </div>

      <div className="card chart-container">
        <h2>Reps Per Day</h2>
        <BarChart data={chartData} />
      </div>

      <div className="card chart-container">
        <h2>RPE Over Time</h2>
        <LineChart data={rpeChartData} />
      </div>
      <button onClick={handleLogout}>Logout</button>
    </div>

  );
};

export default Dashboard;
