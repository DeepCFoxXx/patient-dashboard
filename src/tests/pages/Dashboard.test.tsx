import '@testing-library/jest-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { MemoryRouter, useNavigate } from 'react-router-dom';
import { activityLogs } from '../../data/activityLogs';
import { patientData } from '../../data/patientData';
import {
  prepareChartData,
  prepareRpeData
} from '../../helpers/sessionHelpers';
import Dashboard from '../../pages/Dashboard';

jest.mock('../../components/BarChart', () => ({
  __esModule: true,
  default: ({ data }: { data: Record<string, unknown> }) => (
    <div data-testid="mock-bar-chart">{JSON.stringify(data)}</div>
  ),
}));

jest.mock('../../components/LineChart', () => ({
  __esModule: true,
  default: ({ data }: { data: Record<string, unknown> }) => (
    <div data-testid="mock-line-chart">{JSON.stringify(data)}</div>
  ),
}));

jest.mock('../../components/PieChart', () => ({
  __esModule: true,
  default: ({ data }: { data: Record<string, unknown> }) => (
    <div data-testid="mock-pie-chart">{JSON.stringify(data)}</div>
  ),
}));

jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom');
  return {
    ...actual,
    useNavigate: jest.fn()
  };
});

const renderWithRouter = (
  ui: React.ReactElement,
  { route = '/' } = {}
) => {
  return render(
    <MemoryRouter
      initialEntries={[{ pathname: route, search: '', hash: '', state: {} }]}
      initialIndex={0}
    >
      {ui}
    </MemoryRouter>
  );
};

describe('Dashboard Component', () => {
  it('renders patient details and charts (loaded state)', async () => {
    renderWithRouter(<Dashboard />);
    await waitFor(() =>
      expect(
        screen.getByText(`Name: ${patientData.firstName} ${patientData.lastName}`)
      ).toBeInTheDocument()
    );
    expect(screen.getByText(`Date of Birth: ${patientData.dateOfBirth}`)).toBeInTheDocument();
    expect(screen.getByTestId('mock-pie-chart')).toBeInTheDocument();
  });

  it('calls logout and navigates when Logout is clicked', async () => {
    const navigate = jest.fn();
    const mockedUseNavigate = useNavigate as jest.Mock;
    mockedUseNavigate.mockReturnValue(navigate);

    renderWithRouter(<Dashboard />);
    await waitFor(() =>
      expect(
        screen.getByText(`Name: ${patientData.firstName} ${patientData.lastName}`)
      ).toBeInTheDocument()
    );
    const removeItemSpy = jest.spyOn(Storage.prototype, 'removeItem');
    fireEvent.click(screen.getByText('Logout'));
    expect(removeItemSpy).toHaveBeenCalledWith('token');
    expect(navigate).toHaveBeenCalledWith('/');
  });

  it('matches the snapshot (loaded state)', async () => {
    const { asFragment } = renderWithRouter(<Dashboard />);
    await waitFor(() =>
      expect(
        screen.getByText(`Name: ${patientData.firstName} ${patientData.lastName}`)
      ).toBeInTheDocument()
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('passes correct data to charts', async () => {
    renderWithRouter(<Dashboard />);
    await waitFor(() =>
      expect(screen.getByTestId('mock-bar-chart')).toBeInTheDocument()
    );
    const chartData = prepareChartData(activityLogs);
    const rpeChartData = prepareRpeData(activityLogs);
    expect(screen.getByTestId('mock-bar-chart').textContent).toBe(
      JSON.stringify(chartData)
    );
    expect(screen.getByTestId('mock-line-chart').textContent).toBe(
      JSON.stringify(rpeChartData)
    );
  });
});