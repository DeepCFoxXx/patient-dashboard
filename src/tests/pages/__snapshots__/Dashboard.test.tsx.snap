// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component matches the snapshot (loaded state) 1`] = `
<DocumentFragment>
  <div
    class="dashboard"
  >
    <div
      class="card patient-details"
    >
      <h2
        class="underline"
      >
        Patient Details
      </h2>
      <div
        class="patient-info"
      >
        <div
          class="patient-info-left"
        >
          <p>
            Name: John Doe
          </p>
          <p>
            Date of Birth: 1950-05-05
          </p>
          <p>
            Average Reps Per Session: 609 reps
          </p>
          <p>
            Average Therapy Duration: 1h 54m 52s
          </p>
          <p>
            Average RPE Per Session: 5.39
          </p>
        </div>
        <div
          class="patient-info-right"
        >
          <h3>
            Average Therapy Duration
          </h3>
          <div
            data-testid="mock-pie-chart"
          >
            [{"id":"Short","label":"Short","value":10},{"id":"Medium","label":"Medium","value":30},{"id":"Long","label":"Long","value":20}]
          </div>
        </div>
      </div>
    </div>
    <div
      class="card chart-container"
    >
      <h2>
        Reps Per Day
      </h2>
      <div
        data-testid="mock-bar-chart"
      >
        [{"date":"2025-01-01","totalReps":1741},{"date":"2025-01-02","totalReps":799},{"date":"2025-01-03","totalReps":1162},{"date":"2025-01-04","totalReps":1333},{"date":"2025-01-05","totalReps":1662}]
      </div>
    </div>
    <div
      class="card chart-container"
    >
      <h2>
        RPE Over Time
      </h2>
      <div
        data-testid="mock-line-chart"
      >
        [{"id":"RPE","data":[{"x":"2025-01-01","y":4},{"x":"2025-01-02","y":4},{"x":"2025-01-03","y":3},{"x":"2025-01-04","y":8},{"x":"2025-01-05","y":9}]}]
      </div>
    </div>
    <button>
      Logout
    </button>
  </div>
</DocumentFragment>
`;
